fix(a11y): don't allow passing in null to FocusMonitor.focusVia

crisbeto opened the pull request:
Passing in null to the `FocusMonitor.focusVia` method is currently supported, but it doesn't make sense. These changes make it impossible via typings.

mmalerba commented:
It seems really strange to me to call a method that sounds like its supposed to focus the element and instead it blurs it. Why not just default the origin to 'program' if null is passed? This will yield the same end result that people are seeing today, and in my mind it makes more sense (since I am in fact focusing it programmatically, 'program' makes sense if I fail to specify an explicit origin)

crisbeto commented:
I agree that it seems weird, but I think that the ideal solution would be not to allow `null` to begin with. I think that defaulting it to `program` is weird as well, because I see `focusVia` as "doing the opposite" of `monitor`, but just in this one case (`null`) it doesn't.

mmalerba commented:
I see it more as a version of the native `focus` method that allows the user to override the origin. I think just switching the implementation to the behavior you're suggesting is super confusing. If we want to switch it we should come up with a better name (e.g. maybe `setFocusState`), and deprecate `focusVia`

crisbeto commented:
It might be easier to talk through it and this change isn't particularly urgent. I'll add it to agenda for tomorrow's meeting.

devversion commented:
I agree with what Miles said. I also consider the `focusVia` method as a simple replacement to `HTMLElement.focus` with the ability to overwrite the focus origin. My thinking is that `null` should just fall back to the `default` (which is `program` at the moment). I agree that `null` probably shouldn't have been added in the past, but changing it's meaning, rather causes confusion IMO.

What's the use-case of providing a method for `blur` in the `FocusMonitor`?

crisbeto commented:
I'm not arguing that it's a valid usage, but based on the way we've typed it looks like it's valid so we might as well treat it that way.

devversion commented:
I see what you are trying to achieve. You are saying that it looks like it's a valid scenario. That seems to depend on how people think of it though? or do you think it's made clear somewhere? 

For me personally, since the method is called `focus`, I'd never expect it to be able to blur an element. Also, `null` as as part of the `FocusOrigin` type does not necessarily sound like `blur` (to me personally). Rather it sounds like `use the default origin`, or `origin is unknown` to me.

crisbeto commented:
@mmalerba @devversion I've reworked it based on our discussion.

mmalerba commented on src/cdk/a11y/focus-monitor/focus-monitor.ts:
```diff
@@ -228,28 +228,28 @@ export class FocusMonitor implements OnDestroy {
     if (typeof nativeElement.focus === 'function') {
-      // Cast the element to `any`, because the TS typings don't have the `options` parameter yet.
-      (nativeElement as any).focus(options);
+      // `focus` isn't available on the server.
```
nit: I think it makes sense to leave this comment where it was, since it's explaining the `if`, not the call to `focus`

crisbeto commented:
I've addressed the feedback @mmalerba. I also had to change the signature for a handful of components where we were accepting a `FocusOrigin` and passing it along to `focusVia` directly.

jelbourn commented:
poke @mmalerba

crisbeto commented:
Closing this given how out of date it is and that it'll be a breaking change.
